R/write-docker.R

```diff
@@ -0,0 +1,104 @@
+DEFAULT_RSPM_REPO_ID <-  "1" # cran
+DEFAULT_RSPM <-  "https://packagemanager.rstudio.com"
+
+#' Write a Dockerfile for a vetiver model
+#'
+#' After creating a Plumber file with [vetiver_write_plumber()], use
+#' `vetiver_write_docker()` to create a Dockerfile plus an `renv.lock` file
+#' for a pinned [vetiver_model()].
+#'
+#' @inheritParams vetiver_pr_predict
+#' @param plumber_file A path for your Plumber file, created via
+#' [vetiver_write_plumber()]. Defaults to `plumber.R` in the working directory.
+#' @param path A path to write the Dockerfile and `renv.lock` lockfile,
+#' capturing the model's package dependencies. Defaults to the working directory.
+#'
+#' @return The content of the Dockerfile, invisibly.
+#' @export
+#'
+#' @examples
+#'
+#' library(pins)
+#' tmp_plumber <- tempfile()
+#' b <- board_temp(versioned = TRUE)
+#' cars_lm <- lm(mpg ~ ., data = mtcars)
+#' v <- vetiver_model(cars_lm, "cars_linear")
+#' vetiver_pin_write(b, v)
+#' vetiver_write_plumber(b, "cars_linear", file = tmp_plumber)
+#' vetiver_write_docker(v, tmp_plumber, tempdir())
+#'
+vetiver_write_docker <- function(vetiver_model,
+                                 plumber_file = "plumber.R",
+                                 path = ".") {
+
+
+    r_version <- glue::glue("ARG R_VERSION={getRversion()}")
+
+    pkgs <- unique(c(infra_pkgs, vetiver_model$metadata$required_pkgs))
+    lockfile <- normalizePath(
+        write_renv_lockfile(path = path, pkgs = pkgs),
+        winslash = "/"
+    )
+    sys_reqs <- glue_sys_reqs(pkgs)
+    copy_renv <- glue("COPY {lockfile} /opt/ml/renv.lock")
+    copy_plumber <- glue("COPY {plumber_file} /opt/ml/plumber.R")
+
+
+    ret <- compact(list(
+        "# Generated by the vetiver package; edit with care\n",
+        ## https://github.com/rstudio/plumber/blob/main/Dockerfile:
+        "FROM rstudio/plumber:latest",
+        r_version,
+        "ENV R_VERSION=$R_VERSION\n",
```

Comment:
I think this does not work, actually, to get the correct version of R into the build.

What is my best option here? Build from an image like `rocker/r-ver` and basically copy the guts of the `rstudio/plumber` Dockerfile?

Reply:
Yes. `rstudio/plumber` Dockerfile is very light weight. I don't believe it is too bad to copy over some of the logic. I knowingly provide `plumber`'s Dockerfile with as minimal infrastructure as possible to encourage discovery, but not really permanent usage.

Thoughts:
* You are correct, we should point to `FROM rocker/r-ver:${R_VERSION}` as the parent image
* If we add `remotes` to https://github.com/tidymodels/vetiver/blob/d565de2035cea3a8bd605ca2f7a46d59aa9315b1/R/write-plumber.R#L83, then it'll get auto installed.
	* Do you need `renv` to be installed from GitHub? It might just be easier to add it to the `infra_pkgs`.
* You are already querying package manager, so all system deps should be auto-installed. ☑️ 
* We would need to add an updated entry point for the file:
```dockerfile
EXPOSE 8000
ENTRYPOINT ["R", "-e", "pr <- plumber::plumb(rev(commandArgs())[1]); pr$run(host = '0.0.0.0', port = 8000)"]
```
This is good, as the `rstudio/plumber` ENTRYPOINT was setting the docs to `TRUE`, which would overwrite your custom docs.

Reply:
Thank you so much @schloerke! If you have time to take a look at the new snapshots tests, that would be fantastic.